### PR TITLE
fix: handle derivatives of time-dependent parameters

### LINKED ISSUE #3480: Code generation contains `Differential` when equations contain variable time parameter
Take the following 

```julia
@component function FilteredInput(; name, x0=0, T=0.1)
    params = @parameters begin
      k(t) = x0
      T = T
    end
    vars = @variables begin
      x(t) = k
      dx(t) = 0
      ddx(t)
    end
    systems = []
    eqs = [
      D(x) ~ dx
      D(dx) ~ ddx
      dx ~ (k - x)/T
    ]
    return ODESystem(eqs, t, vars, params; systems, name)
end

@mtkbuild sys = FilteredInput()
ModelingToolkit.observed(sys)
```

gives...

```julia
julia> ModelingToolkit.observed(sys)
3-element Vector{Equation}:
 dx(t) ~ (k(t) - x(t)) / T
 dxˍt(t) ~ (-dx(t) + Differential(t)(k(t))) / T
 ddx(t) ~ dxˍt(t)
```

Note the `Differential(t)(k(t)))` which cannot be computed.

Manifest:
```
[961ee093] ModelingToolkit v9.66.0
```

### PR DESCRIPTION
**Pull request by AayushSabharwal**
Close #3480

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
But also test that the derivative of say a datainterpolation is non-zero still.

**Comment by ChrisRackauckas** [comment]
Was this found to be okay / safe?

**Comment by AayushSabharwal** [comment]
Not completely. We don't always want the derivative of a discrete parameter to be zero. One such example is CoSimulation FMUs. The spec enables getting derivatives of outputs, and sometimes MTK needs this information (e.g. if the output of an FMU is in an equation that pantelides differentiates). I think the solution is to allow `D(p(t)) ~ ...` equations where `p` is a time-dependent parameter and use that in the mapping added to this PR.

**Review by baggepinnen — APPROVED** [review]

File: test/state_selection.jl [hunk 1]
```diff
@@ -2,7 +2,7 @@ using ModelingToolkit, OrdinaryDiffEq, Test
 using ModelingToolkit: t_nounits as t, D_nounits as D
 
 sts = @variables x1(t) x2(t) x3(t) x4(t)
-params = @parameters u1(t) u2(t) u3(t) u4(t)
```
Comment (ChrisRackauckas) [inline test/state_selection.jl:5]: hehe

**Review by ChrisRackauckas — APPROVED** [review]
